### PR TITLE
PersistentVolume last phase transition time

### PR DESCRIPTION
**Pull request by RomanBednar**
#### What type of PR is this?
/kind feature

#### What this PR does / why we need it:

KEP-3762: https://github.com/kubernetes/enhancements/pull/3796

#### Which issue(s) this PR fixes:

Fixes #

#### Special notes for your reviewer:

None.

#### Does this PR introduce a user-facing change?

```release-note
PersistentVolumes have a new LastPhaseTransitionTime field which holds a timestamp of when the volume last transitioned its phase.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by RomanBednar** [comment]
I verified the feature behavior with the following manual test:

1) Create new PV and bind it with PVC (feature disabled)
```
$ kc get pv/pv-1 -o json | jq '.status'
{
  "phase": "Bound"
}
```

2) Verify the new `lastPhaseTransitionTime` field is not present
```
$ kc get pv/pv-1 -o json | jq '.status.lastPhaseTransitionTime'
null
```

3) Enable feature gate

4) Double check new field is absent
```
$ kc get pv/pv-1 -o json | jq '.status.lastPhaseTransitionTime'
null
```

5) Delete PVC to make PV transition to Released
```
$ kc get pvc
NAME    STATUS   VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS   AGE
pvc-1   Bound    pv-1     1Gi        RWO                           3m55s

$ kc delete pvc/pvc-1
persistentvolumeclaim "pvc-1" deleted

$ kc get pv/pv-1 -o json | jq '.status.phase'
"Released"
```

6) Verify the `lastPhaseTransitionTime` is now set
```
$ kc get pv/pv-1 -o json | jq '.status.lastPhaseTransitionTime'
"2023-03-31T12:40:23Z"
```

7) Disable feature gate

8) Remove claim ref to make PV transition to Available
```
$ kc patch pv pv-1 -p '{"spec":{"claimRef": null}}'
persistentvolume/pv-1 patched
```

9) Verify that PV update reset new `lastPhaseTransitionTime` field to nil

```
$ kc get pv/pv-1 -o json | jq '.status.lastPhaseTransitionTime'
null

File: pkg/registry/core/persistentvolume/strategy.go [hunk 1]
```diff
@@ -139,6 +139,8 @@ func (persistentvolumeStatusStrategy) PrepareForUpdate(ctx context.Context, obj,
 	newPv := obj.(*api.PersistentVolume)
 	oldPv := old.(*api.PersistentVolume)
 	newPv.Spec = oldPv.Spec
+
```
Comment (jsafrane) [inline pkg/registry/core/persistentvolume/strategy.go:142]: An alternative approach to update the `LastPhaseTransitionTime` field would be to update it here

```
if old.status.phase != new.status.phase {
   new.status.lastPhaseTransitionTime = now
}
```

So when *user* or some other controller modifies the phase, the timestamp gets updated too. I am leaving it here as an alternative.

File: pkg/features/kube_features.go [hunk 2]
```diff
@@ -624,6 +624,13 @@ const (
 	// Enables PDBUnhealthyPodEvictionPolicy for PodDisruptionBudgets
 	PDBUnhealthyPodEvictionPolicy featuregate.Feature = "PDBUnhealthyPodEvictionPolicy"
 
+	// owner: @RomanBednar
+	// kep: https://kep.k8s.io/3762
+	// alpha: v1.27
```
[the remark this answers is not in the document]
Comment (RomanBednar) [reply pkg/features/kube_features.go:629]: Fixed.

File: pkg/apis/core/validation/validation.go [hunk 3]
```diff
@@ -2011,9 +2014,29 @@ func ValidatePersistentVolumeStatusUpdate(newPv, oldPv *core.PersistentVolume, o
 	if len(newPv.ResourceVersion) == 0 {
 		allErrs = append(allErrs, field.Required(field.NewPath("resourceVersion"), ""))
 	}
+
+	if opts.EnablePersistentVolumeLastPhaseTransitionTime {
+		timestampErrs := validateLastPhaseTransition(&newPv.Status, &oldPv.Status)
+		if len(timestampErrs) > 0 {
+			for _, timestampErr := range timestampErrs {
+				allErrs = append(allErrs, timestampErr)
+			}
```
[the remark this answers is not in the document]
Comment (RomanBednar) [reply pkg/apis/core/validation/validation.go:2023]: Fixed.

File: pkg/registry/core/persistentvolume/strategy.go (hunk 1 above)
Comment (RomanBednar) [reply pkg/registry/core/persistentvolume/strategy.go:142]: Good point, I think we have not considered that in the design initially. Let's see if we get more feedback or find a use case for allowing phase updates from other components or users, we can change this later. To me it makes more sense now to leave this field for pv controller only.

**Comment by jsafrane** [comment]
lgtm-ish from storage POV, waiting for an API review

File: staging/src/k8s.io/api/core/v1/types.go [hunk 4]
```diff
@@ -411,6 +411,9 @@ type PersistentVolumeStatus struct {
 	// for machine parsing and tidy display in the CLI.
 	// +optional
 	Reason string `json:"reason,omitempty" protobuf:"bytes,3,opt,name=reason"`
+	// lastPhaseTransitionTime represents a point in time as a timestamp of when a volume last transitioned its phase.
```
Comment (carlory) [inline staging/src/k8s.io/api/core/v1/types.go:414]:
```suggestion
	// lastPhaseTransitionTime is the time the phase transitioned from one to another.
	// This is an alpha field and requires enabling PersistentVolumeLastPhaseTransitionTime feature.
	// +featureGate=PersistentVolumeLastPhaseTransitionTime
```

File: pkg/apis/core/validation/validation.go [hunk 5]
```diff
@@ -2006,14 +2009,32 @@ func ValidatePersistentVolumeUpdate(newPv, oldPv *core.PersistentVolume, opts Pe
...
 	if len(newPv.ResourceVersion) == 0 {
 		allErrs = append(allErrs, field.Required(field.NewPath("resourceVersion"), ""))
 	}
+
+	if opts.EnablePersistentVolumeLastPhaseTransitionTime {
+		timestampErrs := validateLastPhaseTransition(&newPv.Status, &oldPv.Status)
+		if len(timestampErrs) > 0 {
+			allErrs = append(allErrs, timestampErrs...)
+		}
+	}
```
Comment (carlory) [inline pkg/apis/core/validation/validation.go:2023]:
```suggestion
	if opts.EnablePersistentVolumeLastPhaseTransitionTime {
		allErrs = append(allErrs, validateLastPhaseTransition(&newPv.Status, &oldPv.Status)...)
	}
```

File: pkg/apis/core/validation/validation.go [hunk 6]
```diff
@@ -2006,14 +2009,32 @@ func ValidatePersistentVolumeUpdate(newPv, oldPv *core.PersistentVolume, opts Pe
...
 
+func validateLastPhaseTransition(newPvStatus, oldPvStatus *core.PersistentVolumeStatus) (errs field.ErrorList) {
+	if oldPvStatus.LastPhaseTransitionTime == nil {
+		return
+	}
+	if newPvStatus.LastPhaseTransitionTime != nil && newPvStatus.LastPhaseTransitionTime.Before(oldPvStatus.LastPhaseTransitionTime.DeepCopy()) {
+		errs = append(errs, field.Forbidden(field.NewPath("status", "lastPhaseTransitionTime"), "new timestamp is not valid because it is older than current one"))
+	}
+	return
+}
```
Comment (carlory) [inline pkg/apis/core/validation/validation.go:2036]:
```suggestion
func validateLastPhaseTransition(newPvStatus, oldPvStatus *core.PersistentVolumeStatus) field.ErrorList {
        allErrs := field.ErrorList{}
	if oldPvStatus.LastPhaseTransitionTime.IsZero() {
		return allErrs
	}
	if !newPvStatus.LastPhaseTransitionTime.IsZero() && newPvStatus.LastPhaseTransitionTime.Before(oldPvStatus.LastPhaseTransitionTime) {
		allErrs = append(allErrs, field.Forbidden(field.NewPath("status", "lastPhaseTransitionTime"), "must be newer than before"))
	}
	return allErrs
}
```

**Comment by RomanBednar** [comment]
One test failing with "connection reset by peer" on http request, other one seems like a pod failed to clean up in time - both appear to be transient errors.

/retest

**Comment by carlory** [comment]
It seems that this PR is ready to review. Cloud you take a quick look at this? 

cc @jsafrane @caesarxuchao @jsafrane

**Comment by jsafrane** [comment]
It needs an API review. Let's see if this helps to get it on the radar:
/lgtm

File: pkg/apis/core/validation/validation.go [hunk 7]
```diff
@@ -1681,7 +1684,7 @@ func ValidationOptionsForPersistentVolume(pv, oldPv *core.PersistentVolume) Pers
 	return opts
```
Comment (msau42) [inline pkg/apis/core/validation/validation.go:1684]: In general, once a field is set, we want to continue to validate the field so that it remains valid even if the feature gate is off.

File: pkg/apis/core/validation/validation_test.go [hunk 8]
```diff
@@ -638,7 +638,7 @@ func TestValidatePersistentVolumeSpec(t *testing.T) {
 		},
 	}
 	for name, scenario := range scenarios {
-		opts := PersistentVolumeSpecValidationOptions{}
+		opts := PersistentVolumeValidationOptions{}
```
[the remark this answers is not in the document]
Comment (RomanBednar) [reply pkg/apis/core/validation/validation_test.go:649]: This was just a refactor to have clear naming. Do you mean the `ValidatePersistentVolumeStatusUpdate`? If so then this does not contain anything new if we don't validate the values ([comment](https://github.com/kubernetes/kubernetes/pull/116469#discussion_r1244625032) above).

File: staging/src/k8s.io/api/core/v1/types.go [hunk 9]
```diff
@@ -411,6 +411,11 @@ type PersistentVolumeStatus struct {
 	// for machine parsing and tidy display in the CLI.
 	// +optional
 	Reason string `json:"reason,omitempty" protobuf:"bytes,3,opt,name=reason"`
+	// lastPhaseTransitionTime is the time the phase transitioned from one to another.
+	// This is an alpha field and requires enabling PersistentVolumeLastPhaseTransitionTime feature.
+	// +featureGate=PersistentVolumeLastPhaseTransitionTime
+	// +optional
+	LastPhaseTransitionTime *metav1.Time `json:"lastPhaseTransitionTime,omitempty" protobuf:"bytes,8,opt,name=lastPhaseTransitionTime"`
```
[the remark this answers is not in the document]
Comment (RomanBednar) [reply staging/src/k8s.io/api/core/v1/types.go]: Yes, fixed it.

File: pkg/apis/core/validation/validation.go [hunk 10]
```diff
@@ -2006,11 +2009,27 @@ func ValidatePersistentVolumeUpdate(newPv, oldPv *core.PersistentVolume, opts Pe
...
+
+	return allErrs
+}
+
+func validateLastPhaseTransition(newPvStatus, oldPvStatus *core.PersistentVolumeStatus) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if oldPvStatus.LastPhaseTransitionTime.IsZero() || newPvStatus.LastPhaseTransitionTime.IsZero() {
+		return allErrs
+	}
+	if newPvStatus.LastPhaseTransitionTime.Before(oldPvStatus.LastPhaseTransitionTime) {
```
[the remark this answers is not in the document]
Comment (RomanBednar) [reply pkg/apis/core/validation/validation.go]: I see the point, removing - the field is not restricted/validated now.

File: pkg/api/persistentvolume/util.go [hunk 11]
```diff
@@ -31,16 +31,26 @@ const (
...
 		if pvSpec.CSI != nil {
 			pvSpec.CSI.NodeExpandSecretRef = nil
 		}
 	}
 }
 
+// DropDisabledStatusFields removes disabled fields from the pv status.
+// This should be called from PrepareForCreate/PrepareForUpdate for all resources containing a pv status.
+func DropDisabledStatusFields(pvStatus *api.PersistentVolumeStatus) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.PersistentVolumeLastPhaseTransitionTime) {
```
[the remark this answers is not in the document]
Comment (RomanBednar) [reply pkg/api/persistentvolume/util.go]: If I understand correctly, doing so will basically remove the option to disable the feature? Once the field is set it will never be removed regardless of the feature gate state, is that what we want?

File: pkg/api/persistentvolume/util.go [hunk 12]
```diff
@@ -31,16 +31,24 @@ const (
...
 		if pvSpec.CSI != nil {
 			pvSpec.CSI.NodeExpandSecretRef = nil
 		}
 	}
 }
 
+// DropDisabledStatusFields removes disabled fields from the pv status.
+// This should be called from PrepareForCreate/PrepareForUpdate for all resources containing a pv status.
+func DropDisabledStatusFields(pvStatus *api.PersistentVolumeStatus) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.PersistentVolumeLastPhaseTransitionTime) && pvStatus.LastPhaseTransitionTime.IsZero() {
```
Comment (carlory) [inline pkg/api/persistentvolume/util.go:47]: Why we need check `pvStatus.LastPhaseTransitionTime.IsZero()`?

File: pkg/apis/core/validation/validation.go [hunk 13]
```diff
@@ -2006,11 +2009,27 @@ func ValidatePersistentVolumeUpdate(newPv, oldPv *core.PersistentVolume, opts Pe
...
+	if opts.EnablePersistentVolumeLastPhaseTransitionTime {
+		allErrs = append(allErrs, validateLastPhaseTransition(&newPv.Status, &oldPv.Status)...)
+	}
+
+	return allErrs
+}
+
+func validateLastPhaseTransition(newPvStatus, oldPvStatus *core.PersistentVolumeStatus) field.ErrorList {
+	allErrs := field.ErrorList{}
+	if oldPvStatus.LastPhaseTransitionTime.IsZero() || newPvStatus.LastPhaseTransitionTime.IsZero() {
```
[the remark this answers is not in the document]
Comment (carlory) [reply pkg/apis/core/validation/validation.go]: According to [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-storage/3762-persistent-volume-last-phase-transition-time/README.md#design-details), it should remove the timestamp in LastPhaseTransitionTime field during volume status update when feature gate is disabled. 

If it's allowed, we need add feature-gate check in validation. Otherwise, kep should be updated.

File: pkg/apis/core/validation/validation.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (carlory) [reply pkg/apis/core/validation/validation.go]: if the field is not restricted/validated now, we need remove this sentence from kep.

```
validation should not allow updating to a point in time before the current timestamp
```

File: pkg/apis/core/validation/validation.go (hunk 13 above)
[the remark this answers is not in the document]
Comment (RomanBednar) [reply pkg/apis/core/validation/validation.go]: @carlory Opened KEP update PR to match it with changes discussed here: https://github.com/kubernetes/enhancements/pull/4125

**Comment by jsafrane** [comment]
/lgtm
from storage POV

File: pkg/registry/core/persistentvolume/storage/storage_test.go [hunk 14]
```diff
@@ -201,6 +204,65 @@ func TestUpdateStatus(t *testing.T) {
 	}
 }
 
+func TestUpdateStatusLastPhaseTransitionTime(t *testing.T) {
```
Comment (msau42) [inline pkg/registry/core/persistentvolume/storage/storage_test.go:207]: I think it would be good to make this table driven so you can test the following scenarios:

- Feature gate enabled, phase changes
- Feature gate enabled, phase doesn't change
- Feature gate disabled, old unset, phase changes
- Feature gate disabled, old unset, phase doesn't chase
- Feature gate disabled, old set, phase changes
- Feature gate disabled, old set, phase doesn't change

And in all the scenarios, have the input status set the field (so we can see it being overridden)

File: pkg/registry/core/persistentvolume/strategy.go [hunk 15]
```diff
@@ -139,6 +142,15 @@ func (persistentvolumeStatusStrategy) PrepareForUpdate(ctx context.Context, obj,
...
+	if utilfeature.DefaultFeatureGate.Enabled(features.PersistentVolumeLastPhaseTransitionTime) {
```
[the remark this answers is not in the document]
Comment (msau42) [reply pkg/registry/core/persistentvolume/strategy.go:146]: Regarding who should set the LastTransitionTime, it looks like for CSR, the api server only sets it if it is zero, in order to accomodate older clients: https://github.com/kubernetes/kubernetes/blob/master/pkg/registry/certificates/certificates/strategy.go/#L215

File: pkg/api/persistentvolume/util.go [hunk 16]
```diff
@@ -31,16 +31,24 @@ const (
...
 	if !utilfeature.DefaultFeatureGate.Enabled(features.CSINodeExpandSecret) && !hasNodeExpansionSecrets(oldPVSpec) {
 		if pvSpec.CSI != nil {
 			pvSpec.CSI.NodeExpandSecretRef = nil
 		}
 	}
 }
 
+// DropDisabledStatusFields removes disabled fields from the pv status.
+// This should be called from PrepareForCreate/PrepareForUpdate for all resources containing a pv status.
+func DropDisabledStatusFields(pvStatus *api.PersistentVolumeStatus) {
```
[the remark this answers is not in the document]
Comment (RomanBednar) [reply pkg/api/persistentvolume/util.go]: Done.

File: pkg/apis/core/validation/validation.go [hunk 17]
```diff
@@ -1651,10 +1651,12 @@ var allowedTemplateObjectMetaFields = map[string]bool{
...
 }
 
-// PersistentVolumeSpecValidationOptions contains the different settings for PeristentVolume validation
-type PersistentVolumeSpecValidationOptions struct {
+// PersistentVolumeValidationOptions contains the different settings for PeristentVolume validation
+type PersistentVolumeValidationOptions struct {
 	// Allow spec to contain the "ReadWiteOncePod" access mode
 	AllowReadWriteOncePod bool
+	// Allow updating pv last phase change timestamp
+	EnablePersistentVolumeLastPhaseTransitionTime bool
```
[the remark this answers is not in the document]
Comment (RomanBednar) [reply pkg/apis/core/validation/validation.go]: Removed.

File: pkg/registry/core/persistentvolume/strategy.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (RomanBednar) [reply pkg/registry/core/persistentvolume/strategy.go:153]: I refreshed and read your comments after updating the PR...should I revert it back then?

File: pkg/api/persistentvolume/util.go [hunk 18]
```diff
@@ -31,16 +31,28 @@ const (
...
 	}
 }
 
+// DropDisabledStatusFields removes disabled fields from the pv status.
+// This should be called from PrepareForUpdate for all resources containing a pv status.
+func DropDisabledStatusFields(oldStatus, newStatus *api.PersistentVolumeStatus) {
+	if !utilfeature.DefaultFeatureGate.Enabled(features.PersistentVolumeLastPhaseTransitionTime) && oldStatus.LastPhaseTransitionTime.IsZero() {
+		newStatus.LastPhaseTransitionTime = nil
+	} else {
+		if oldStatus.LastPhaseTransitionTime != nil {
```
Comment (msau42) [inline pkg/api/persistentvolume/util.go:50]: This means that we will never allow changing the time once it's set regardless of feature gate.

I think just disallowing initial setting if the feature gate is off is sufficient.

File: pkg/registry/core/persistentvolume/storage/storage_test.go [hunk 19]
```diff
@@ -208,3 +211,105 @@ func TestShortNames(t *testing.T) {
...
+				t.Errorf("unexpected error: %v", err)
+			}
+			pvOut := obj.(*api.PersistentVolume)
+
+			if tc.timestampExpected {
+				if pvOut.Status.LastPhaseTransitionTime == nil {
+					t.Fatalf("Expected PV %s to have a timestamp but got nil", pvOut.GetName())
+				}
+			} else {
+				if pvOut.Status.LastPhaseTransitionTime != nil {
```
Comment (msau42) [inline pkg/registry/core/persistentvolume/storage/storage_test.go:309]: I think it would be better to validate the exact timestamp. That way we can test whether or not we keep what the user sets in each scenario.

Take a look at the CSR example for how to stub out the `now` method for unit testing.

File: pkg/registry/core/persistentvolume/strategy.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (msau42) [reply pkg/registry/core/persistentvolume/strategy.go:153]: The behavior @jsafrane described sounds good to me. It would be good to document the auto-setting behavior in the api comments.

File: pkg/registry/core/persistentvolume/storage/storage_test.go (hunk 19 above)
Comment (RomanBednar) [reply pkg/registry/core/persistentvolume/storage/storage_test.go]: Ack, added.

File: pkg/registry/core/persistentvolume/strategy.go (hunk 1 above)
Comment (RomanBednar) [reply pkg/registry/core/persistentvolume/strategy.go:148]: Done - moved to strategy.

File: staging/src/k8s.io/api/core/v1/types.go [hunk 20]
```diff
@@ -411,6 +411,11 @@ type PersistentVolumeStatus struct {
 	// for machine parsing and tidy display in the CLI.
 	// +optional
 	Reason string `json:"reason,omitempty" protobuf:"bytes,3,opt,name=reason"`
+	// lastPhaseTransitionTime is the time the phase transitioned from one to another.
+	// This is an alpha field and requires enabling PersistentVolumeLastPhaseTransitionTime feature.
```
Comment (msau42) [inline staging/src/k8s.io/api/core/v1/types.go:415]: Also copy the comment changes to here. This file is the one that is used to generate the API docs.

**Comment by msau42** [comment]
/approve

Just have one nit. Will let @jsafrane lgtm once addressed.

**Comment by jpbetz** [comment]
/approve
for applyconfiguration changes

File: test/e2e_node/resource_metrics_test.go [hunk 21]
```diff
@@ -106,7 +106,15 @@ var _ = SIGDescribe("ResourceMetricsAPI [NodeFeature:ResourceMetrics]", func() {
...
-			})
+
+				"pod_swap_usage_bytes": gstruct.MatchElements(podID, gstruct.IgnoreExtras, gstruct.Elements{
+					fmt.Sprintf("%s::%s", f.Namespace.Name, pod0): boundedSample(0*e2evolume.Kb, 80*e2evolume.Mb),
+					fmt.Sprintf("%s::%s", f.Namespace.Name, pod1): boundedSample(0*e2evolume.Kb, 80*e2evolume.Mb),
+				}),
+			}),
+				haveKeys("scrape_error", "node_cpu_usage_seconds_total", "node_memory_working_set_bytes", "container_cpu_usage_seconds_total",
+					"container_memory_working_set_bytes", "container_start_time_seconds", "pod_cpu_usage_seconds_total", "pod_memory_working_set_bytes"),
+			)
```
Comment (jsafrane) [inline test/e2e_node/resource_metrics_test.go:117]: this does not belong to this PR, IMO

File: staging/src/k8s.io/api/core/v1/types.go (hunk 20 above)
Comment (RomanBednar) [reply staging/src/k8s.io/api/core/v1/types.go:416]: Done - copied the comment so they match for private and public API.

File: pkg/apis/core/validation/validation.go [hunk 22]
```diff
@@ -1651,8 +1651,8 @@ var allowedTemplateObjectMetaFields = map[string]bool{
 	"Labels":      true,
 }
 
-// PersistentVolumeSpecValidationOptions contains the different settings for PeristentVolume validation
-type PersistentVolumeSpecValidationOptions struct {
+// PersistentVolumeValidationOptions contains the different settings for PeristentVolume validation
+type PersistentVolumeValidationOptions struct {
```
Comment (liggitt) [inline pkg/apis/core/validation/validation.go:1655]: why was this type renamed? the options in it only apply to spec, right?

File: pkg/features/kube_features.go [hunk 23]
```diff
@@ -1069,6 +1076,8 @@ var defaultKubernetesFeatureGates = map[featuregate.Feature]featuregate.FeatureS
 
 	KubeProxyDrainingTerminatingNodes: {Default: false, PreRelease: featuregate.Alpha},
 
+	PersistentVolumeLastPhaseTransitionTime: {Default: false, PreRelease: featuregate.Alpha},
```
Comment (liggitt) [inline pkg/features/kube_features.go:1079]: nit: put this in alphabetical order with the other feature gates

File: pkg/registry/core/persistentvolume/strategy.go [hunk 24]
```diff
@@ -134,11 +137,22 @@ func (persistentvolumeStatusStrategy) GetResetFields() map[fieldpath.APIVersion]
...
+
 // PrepareForUpdate sets the Spec field which is not allowed to be changed when updating a PV's Status
 func (persistentvolumeStatusStrategy) PrepareForUpdate(ctx context.Context, obj, old runtime.Object) {
+	now := nowFunc()
 	newPv := obj.(*api.PersistentVolume)
 	oldPv := old.(*api.PersistentVolume)
 	newPv.Spec = oldPv.Spec
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.PersistentVolumeLastPhaseTransitionTime) {
+		if oldPv.Status.Phase != newPv.Status.Phase {
```
Comment (liggitt) [inline pkg/registry/core/persistentvolume/strategy.go:150]: It's surprising to stomp a client-set value unconditionally.

I expected the server to only populate the field if the client was 1) clearing it on update or 2) not changing it when the phase changed. 

Something like this:

```go
switch {
case oldPv.Status.Phase == newPv.Status.Phase && newPv.Status.LastPhaseTransitionTime == nil:
  // phase didn't change, preserve the existing transition time if set
  newPv.Status.LastPhaseTransitionTime = oldPv.Status.LastPhaseTransitionTime

case oldPv.Status.Phase != newPv.Status.Phase && (newPv.Status.LastPhaseTransitionTime == nil || newPv.Status.LastPhaseTransitionTime.Equal(oldPv.Status.LastPhaseTransitionTime)):
  // phase changed and client didn't set or didn't change the transition time
  now := nowFunc()
  newPv.Status.LastPhaseTransitionTime = &now
}
```

Comment (liggitt) [reply pkg/registry/core/persistentvolume/strategy.go:150]: is the field expected to be populated on create as well if empty?

Comment (jsafrane) [reply pkg/registry/core/persistentvolume/strategy.go:150]: We had long discussion with @msau42 who / when should set the field and we came up with the current code.
Clients can set it to whatever they want, but the API server overwrites it if and only if the phase changes.

Comment (jsafrane) [reply pkg/registry/core/persistentvolume/strategy.go:150]: one of good arguments was that the timestamp will be close to audit events

Comment (liggitt) [reply pkg/registry/core/persistentvolume/strategy.go:150]: why is this field special in the server overwriting the client-set time, when compared to other things like pod timestamps set by the kubelet, condition timestamps set by the controller, etc?

Comment (jsafrane) [reply pkg/registry/core/persistentvolume/strategy.go:150]: It's not that special. So shall we move changing the field back to PV controller, i.e. slip 1.28?

Comment (liggitt) [reply pkg/registry/core/persistentvolume/strategy.go:150]: I think the suggestion in https://github.com/kubernetes/kubernetes/pull/116469#discussion_r1270744169 would still populate and update the field for a controller that did not set the timestamp

File: pkg/apis/core/validation/validation.go (hunk 22 above)
Comment (RomanBednar) [reply pkg/apis/core/validation/validation.go]: The intention was to validate the new status field as well, but we dropped it along the way and this commit got forgotten - dropping it as well as the rename is not needed now.

File: pkg/registry/core/persistentvolume/strategy.go (hunk 24 above)
Comment (RomanBednar) [reply pkg/registry/core/persistentvolume/strategy.go]: @liggitt I've updated the PR to reflect this behavior. Only the 3rd test case changed so that the client provided value is preserved, the rest is the same - I suppose that's the expected behavior now?

File: pkg/registry/core/persistentvolume/strategy_test.go [hunk 25]
```diff
@@ -34,3 +41,274 @@ func TestSelectableFieldLabelConversions(t *testing.T) {
...
+					Name: "foo",
+				},
+				Status: api.PersistentVolumeStatus{
+					Phase:                   api.VolumePending,
+					LastPhaseTransitionTime: &later,
+				},
+			},
+		},
+		{
+			name: "feature enabled: user timestamp is respected on phase change",
```
Comment (msau42) [inline pkg/registry/core/persistentvolume/strategy_test.go:120]: Can you add one more test case for the timestamp being updated on phase change but the user didn't update the timestamps?

Comment (RomanBednar) [reply pkg/registry/core/persistentvolume/strategy_test.go:150]: @msau42 Added.

**Comment by liggitt** [comment]
/approve

for strategy bit
@msau42 can lgtm

File: pkg/registry/core/persistentvolume/strategy.go (hunk 24 above)
Comment (msau42) [reply pkg/registry/core/persistentvolume/strategy.go]: There's one more question on if the timestamp should be initialized on create as well. I think it would be nice to have so that any tooling that is looking at this only needs to look at one field instead of special casing the creation case.

Comment (liggitt) [reply pkg/registry/core/persistentvolume/strategy.go]: Create appears to be forcing an empty status and phase, so leaving the time stamp empty seems ok?

Comment (msau42) [reply pkg/registry/core/persistentvolume/strategy.go]: Remind me again when does defaulting happen in relation to `PrepareForCreate()` and `Validate()`. We have logic to default the phase to "Pending": https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/core/v1/defaults.go#L300

Comment (liggitt) [reply pkg/registry/core/persistentvolume/strategy.go]: Huh... defaulting happens on the incoming object before PrepareForCreate, so clearing status will wipe out the default

When the newly created object is next read from etcd, defaulting will fill the status.phase field again. If you want the newly created object to have phase="Pending" and lastTransitionTimestamp=now(), that seems ok to set in PrepareForCreate()

File: pkg/registry/core/persistentvolume/strategy.go [hunk 26]
```diff
@@ -66,7 +69,13 @@ func (persistentvolumeStrategy) PrepareForCreate(ctx context.Context, obj runtim
 	pv := obj.(*api.PersistentVolume)
 	pv.Status = api.PersistentVolumeStatus{}
 
-	pvutil.DropDisabledFields(&pv.Spec, nil)
+	if utilfeature.DefaultFeatureGate.Enabled(features.PersistentVolumeLastPhaseTransitionTime) {
```
Comment (msau42) [inline pkg/registry/core/persistentvolume/strategy.go:72]: Can you add a test for this

File: pkg/registry/core/persistentvolume/strategy_test.go (hunk 25 above)
Comment (msau42) [reply pkg/registry/core/persistentvolume/strategy_test.go:150]: I didn't see the new test case. I am thinking of the case where both old.LastPhaseTransitionTime and new.LastPhaseTransitionTime are the same, but the phase changes.

File: pkg/registry/core/persistentvolume/strategy.go (hunk 26 above)
Comment (RomanBednar) [reply pkg/registry/core/persistentvolume/strategy.go:72]: Added.

File: pkg/registry/core/persistentvolume/strategy_test.go (hunk 25 above)
Comment (RomanBednar) [reply pkg/registry/core/persistentvolume/strategy_test.go:150]: Ack, that's a different test case, adding one more.

File: pkg/registry/core/persistentvolume/strategy_test.go [hunk 27]
```diff
@@ -34,3 +41,398 @@ func TestSelectableFieldLabelConversions(t *testing.T) {
...
+				},
+				Status: api.PersistentVolumeStatus{
+					Phase:                   api.VolumePending,
+					LastPhaseTransitionTime: &now,
+				},
+			},
+		},
+		{
+			name: "feature disabled: pv does not have phase and timestamp",
+			fg:   true,
```
Comment (msau42) [inline pkg/registry/core/persistentvolume/strategy_test.go:409]: This should be flipped to false, and the status should be empty

Comment (RomanBednar) [reply pkg/registry/core/persistentvolume/strategy_test.go]: Corrected.

**Comment by msau42** [comment]
1.28 exception approval: https://kubernetes.slack.com/archives/C2C40FMNF/p1689859135443509
